### PR TITLE
:sparkles: add a new field that reports errors in case resource was incompatible

### LINKED ISSUE #2911: bug: Report detailed compatibility errors on incompatible `SyncTarget.Status.ResourcesToSync`
### Describe the bug

When a resource to sync on a SyncTarget is incompatible, there are not details about the compatibility errors between the KCP schema and the SyncTarget schema.
That makes it quite hard to understand what happens.

### Steps To Reproduce

1. Create a SyncTarget that binds to the `root:compute:kubernetes` APIExport
2. Install this SyncTarget in a Kubernetes physical cluster with an old release (1.21 for example)
3. See that some of the resources in the `SyncTarget.Status.ResourcesToSync` list have the `Incompatible` state, but don't provide any more detail.

### Expected Behaviour

We should have a compatibility detailed error message.
This should be possible, since the internal code used to do the compatibility check returns a list of errors (missing field, incompatible types, etc...)

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by rahulii**
## Summary
This PR add a new field that reports incompatibility errors!
## Related issue(s)

Fixes kcp-dev/contrib-tmc#33

### CODE REVIEW

File: pkg/reconciler/workload/synctargetexports/synctargetcompatible_reconcile.go [hunk 1]
```diff
@@ -123,6 +123,7 @@ func (e *apiCompatibleReconciler) reconcile(ctx context.Context, syncTarget *wor
 				field.NewPath(gvr.String()), upstreamSchema, downStreamSchema, false)
 			if err != nil {
 				syncTarget.Status.SyncedResources[i].State = workloadv1alpha1.ResourceSchemaIncompatibleState
+				syncTarget.Status.SyncedResources[i].Message = err.Error()
```
Comment (davidfestal) [inline pkg/reconciler/workload/synctargetexports/synctargetcompatible_reconcile.go:126]: Do you reset this field to empty when the incompatibility if fixed ?

File: sdk/apis/workload/v1alpha1/synctarget_types.go [hunk 2]
```diff
@@ -136,6 +136,11 @@ type ResourceToSync struct {
 	// +kubebuilder:default=Pending
 	// +optional
 	State ResourceCompatibleState `json:"state,omitempty"`
+	
+	// A human readable message indicating details about the compatibility errors
```
Comment (davidfestal) [inline sdk/apis/workload/v1alpha1/synctarget_types.go:140]:
```suggestion
	// A human readable message indicating details about the compatibility errors.
```

File: pkg/reconciler/workload/synctargetexports/synctargetcompatible_reconcile.go (hunk 1 above)
Comment (rahulii) [reply pkg/reconciler/workload/synctargetexports/synctargetcompatible_reconcile.go:126]: made changes!can you PTAL now?

File: pkg/reconciler/workload/synctargetexports/synctargetcompatible_reconcile.go [hunk 3]
```diff
@@ -123,11 +123,15 @@ func (e *apiCompatibleReconciler) reconcile(ctx context.Context, syncTarget *wor
...
 				continue
 			}
 
 			// since version is ordered, so if the current version is comptaible, we can skip the check on other versions.
 			syncTarget.Status.SyncedResources[i].State = workloadv1alpha1.ResourceSchemaAcceptedState
+			if syncTarget.Status.SyncedResources[i].Message != "" {
```
Comment (stevekuznetsov) [inline pkg/reconciler/workload/synctargetexports/synctargetcompatible_reconcile.go:132]: nit: no need for the `if` check here, can unconditionally clear it

**Review by stevekuznetsov — COMMENTED** [review]
/ok-to-test

**Comment by stevekuznetsov** [comment]
If an end-to-end test already exists for this case, can we check that message is set how we expect there?

**Comment by rahulii** [comment]
> If an end-to-end test already exists for this case, can we check that message is set how we expect there?

@stevekuznetsov where are the e2e tests located?

**Comment by davidfestal** [comment]
@rahulii You should be able to add the comparison of the message in the following e2e test:

https://github.com/kcp-dev/kcp/blob/24b8050b6540b5a2913b6db72b7db3878cd5df2f/test/e2e/reconciler/locationworkspace/synctarget_test.go#L118-L124

**Comment by mjudeikis** [comment]
/close
TMC related

**Pull request closed without merging**